lib/bolt/inventory/inventory2.rb

```diff
@@ -219,102 +197,202 @@ def expand_targets(targets)
+        data['config'] ||= {}
+        data['vars'] ||= {}
+        data['facts'] ||= {}
+        data['features'] = data['features'] ? Set.new(data['features']) : Set.new
+        data['groups'] ||= []
+        data['options'] ||= {}
+        data['plugin_hooks'] ||= {}
+        data['target_alias'] ||= []
+
+        # Every call to update_target will rebuild this state based on merging together target, group, and config data
+        data['cached_state'] = {}
+
+        target = Target2.new(nil, data['name'])
+        target.inventory = self
+        data['self'] = target
+        @targets[data['name']] = data
+        target
       end
-      private :set_facts
+      private :create_target
+
+      def create_target_from_plan(data)
+        t_name = data['name'] || data['uri']
 
-      def set_plugin_hooks(tname, hash)
-        if hash
-          @target_plugin_hooks[tname] ||= {}
-          @target_plugin_hooks[tname].merge!(hash)
+        # If target allready exists, delete old and replace with new
+        # Otherwise add to new to all group
+        if @targets[t_name]
+          @targets.delete(t_name)
+          t = create_target(t_name, data)
+          update_target(t)
+        else
+          t = create_target(t_name, data)
+          update_target(t)
+          add_to_group([t], 'all')
         end
+        t
       end
-      private :set_plugin_hooks
 
-      def add_target(current_group, target, desired_group, track = { 'all' => nil })
-        if current_group.name == desired_group
-          # Group to add to is found
-          t_name = target.name
-          # Add target to targets hash
-          target_hash = { 'name' => t_name }.merge(target.options)
-          target_hash['uri'] = target.uri if target.uri
-          current_group.targets[t_name] = target_hash
-
-          # Inherit facts, vars, and features from hierarchy
-          current_group_data = { facts: current_group.facts,
-                                 vars: current_group.vars,
-                                 features: current_group.features,
-                                 plugin_hooks: current_group.plugin_hooks }
-          data = inherit_data(track, current_group.name, current_group_data)
-          set_facts(t_name, @target_facts[t_name] ? data[:facts].merge(@target_facts[t_name]) : data[:facts])
-          set_vars_from_hash(t_name, @target_vars[t_name] ? data[:vars].merge(@target_vars[t_name]) : data[:vars])
-          data[:features].each do |feature|
-            set_feature(target, feature)
-          end
-          hook_data = @config.plugin_hooks.merge(data[:plugin_hooks])
-          hash = if @target_plugin_hooks[t_name]
-                   hook_data.merge(@target_plugin_hooks[t_name])
-                 else
-                   hook_data
-                 end
-          set_plugin_hooks(t_name, hash)
-          return true
+      # CODEREVIEW: This is helpful in group2 for validation. Where is a good place for it?
```

Comment:
Honestly `Bolt::Util` seems like a good place for it? It's a helper method, and since it's not really part of the groups or inventory's functionality I think util makes the most sense.

Reply:
I dont think it belongs in util. There is some specific logic to ensure that the `Uri` object that is returned is compatible with how we use it in inventory (for example, how we append `//` conditionally and forbid empty strings). Maybe it should go in `Bolt::Inventory` becuase both `Inventory2 and `Group2` are derived from that?

Reply:
Inventory2 and Group2 are just nested classes not child classes, so they wouldn't inherit the method. If we don't want to use it elsewhere it makes the most sense in Inventory2 - it isn't used in Inventory, and conceptually fits with being an inventory2 helper method.

Reply:
@puppetlabs/bolt Open question